document ResourceUsage and Metrics CR (#975)
kubernetes-sigs/kwok: site/content/en/docs/user/resource-usage-configuration.md

  namespace: <string>
spec:
  usages:
  - containers:
    - <string>
    usage:
      cpu:
        value: <quantity>
        expression: <string>
      memory:
        value: <quantity>
        expression: <string>
```

To associate a ResourceUsage with a certain pod to be simulated, users must ensure `metadata.name` and `metadata.namespace` are inconsistent with the name and namespace of the target pod.

The resource usages of a pod are specified via `usages` field.
The `usages` field are organized by groups, with each corresponding to a collection of containers that shares a same resource usage simulation setting.
Each group consists of a list of container names (`containers`) and the shared resource usage setting (`usage`).

{{< hint "info" >}}
If `containers` is not given in a group, the `usage` in that group will be applied to all containers of the target pod.
{{< /hint >}}

You can simply set a static [Quantity value] (`100Mi`, `1000m`, etc.) via `cpu.value` and `memory.value` to define the cpu and memory resource usage respectively.
Besides, users are also allowed to provide a [CEL expression] via `expressions` to describe the resource usage more flexibly. For example,

> the following expression tries to extract the cpu resource usage from the pod's annotation if it has or use a default value.

Reviewer: ```suggestion
the following expression tries to extract the cpu resource usage from the pod's annotation if it has or use a default value.

```
Added a blank line

```yaml
expression: |
  "kwok.x-k8s.io/usage-cpu" in pod.metadata.annotations
  ? Quantity(pod.metadata.annotations["kwok.x-k8s.io/usage-cpu"])
  : Quantity("1m")
```

{{< hint "info" >}}
1. `value` has higher priority than `expressions` if both are set.
2. Quantity value must be explicitly wrapped by `Quantity` function in CEL expressions.
{{< /hint >}}

> With CEL expressions, it is even possible to simulate resource usages dynamically. For example, the following expression
> yields CPU usage that grows linearly with time.

Reviewer 2: ```suggestion
With CEL expressions, it is even possible to simulate resource usages dynamically.

For example, the following expression yields memory usage that grows linearly with time.
```

`Quantity("1Mi")` means memory quantity, `Quantity("1m")` is for the CPU.

```yaml
expression: (pod.SinceSecond() / 60.0) * Quantity("1Mi")
```

> Please refer to [built-in CEL extension functions] for an exhausted list that may helpful to config dynamic resource usages.

Reviewer: ```suggestion

Please refer to [built-in CEL extension functions] for an exhausted list that may be helpful to configure dynamic resource usage.
```

ClusterResourceUsage has an additional `selector` field for specifying the target pods to be simulated.
`matchNamespaces` and `matchNames` are both represented as list，which are designed to take pod collections by different levels:

1. If `matchNamespaces` is empty, ClusterResourceUsage will be applied to all pods that are managed by `kwok` and whose names listed in `matchNames`.
2. If `matchNames` is empty, ClusterResourceUsage will be applied to all pods managed by `kwok` and under namespaces listed in `matchNamespaces`.
3. If `matchNames` and `matchNamespaces` are both unset, ClusterResourceUsage will be applied to all pods that `kwok` manages.

The `usages` field of ClusterResourceUsage has the same semantic with the one in ResourceUsage.

Please refer to [pod resource usage from annotation] for a concrete example.

## Where to get the simulation data?

The resource usages defined in ResourceUsage and ClusterResourceUsage resources can be fetched from the metric service of `kwok` at port `10247` with path `/metrics/nodes/{nodeName}/metrics/resource`,
where `{nodeName}` is the name of the fake node that the pod is scheduled to.

> The returned metrics look the same as the response of kubelet's `/metrics/resource` endpoint.

Reviewer: ```suggestion
The returned metrics are similar to the response from kubelet's `/metrics/resource` endpoint.
```

> ---

Reviewer 2: This page is actually similar to Exec, Logs, Attach, PortForward. Let's make the content and style of the descriptions consistent across multiple pages.
